ibtk/src/solvers/impls/PETScLevelSolver.cpp

```diff
@@ -307,6 +314,19 @@ void PETScLevelSolver::deallocateSolverState()
     return;
 } // deallocateSolverState
 
+void PETScLevelSolver::addLinearOperator(Mat& op, MatStructure nonzero_pattern)
+{
+#if !defined(NDEBUG)
+	TBOX_ASSERT(!d_is_initialized);
+	TBOX_ASSERT(op);
+#endif
+
+	d_petsc_extern_mat = op;
+	d_extern_mat_nz_pattern = nonzero_pattern;
+
+	return;
+}// addLinearOperator
+
 /////////////////////////////// PROTECTED ////////////////////////////////////
 
 void PETScLevelSolver::init(Pointer<Database> input_db, const std::string& default_options_prefix)
```

Comment:
> On May 2, 2015, at 2:56 PM, Amneet Bhalla <notifications@github.com> wrote:
> 
> In ibtk/src/solvers/impls/PETScLevelSolver.cpp <https://github.com/IBAMR/IBAMR/pull/6#discussion_r29549789>:
> 
> > @@ -307,6 +314,19 @@ void PETScLevelSolver::deallocateSolverState()
> >      return;
> >  } // deallocateSolverState
> >  
> > +void PETScLevelSolver::addLinearOperator(Mat& op, MatStructure nonzero_pattern)
> > +{
> > +#if !defined(NDEBUG)
> > +	TBOX_ASSERT(!d_is_initialized);
> > +	TBOX_ASSERT(op);
> > +#endif
> > +
> > +	d_petsc_extern_mat = op;
> > +	d_extern_mat_nz_pattern = nonzero_pattern;
> > +
> > +	return;
> > +}// addLinearOperator
> > +
> >  /////////////////////////////// PROTECTED ////////////////////////////////////
> >  
> >  void PETScLevelSolver::init(Pointer<Database> input_db, const std::string& default_options_prefix)
> How about supporting setAdditive{Multiplicative}LinearOperator in the class? The intention of these 'extraneous' Ops will be do some extra work on top of the main linear operator employed in the KSP, e.g., adding diagonal matrix to improve the conditioning of the existing matrix. This will avoid having to create many different classes that differ from the classical KSP matrix in some respect...
> 

I'm not convinced that this is something that belongs in this interface.

If you need to modify the operator following construction, you can always grab the KSP and then modify the operator yourself. If you want to modify the operator that is constructed by a solver, it seems like the right place to do this is in a subclass.